hknweb/candidate/utils.py

```diff
@@ -0,0 +1,181 @@
+
+    bitbyte_link = request.build_absolute_uri(
+        reverse("candidate:bitbyte"))
+    html_content = render_to_string(
+        'candidate/bitbyte_confirm_email.html',
+        {
+            'subject': subject,
+            'confirmed': confirmed,
+            'participants': bitbyte.participants.all(),
+            'bitbyte_link': bitbyte_link,
+            'img_link': get_rand_photo(),
+        }
+    )
+    msg = EmailMultiAlternatives(subject, subject,
+                settings.NO_REPLY_EMAIL, participant_emails)
+    msg.attach_alternative(html_content, "text/html")
+    msg.send()
+
+
+""" What the event types are called on admin site.
+    Code will not work if they're called something else!! """
+map_event_vars = {
+    settings.MANDATORY_EVENT: 'Mandatory',
+    settings.FUN_EVENT: 'Fun',
+    settings.BIG_FUN_EVENT: 'Big Fun',
+    settings.SERV_EVENT: 'Serv',
+    settings.PRODEV_EVENT: 'Prodev',
+    settings.HANGOUT_EVENT: 'Hangout',
+    settings.BITBYTE_ACTIVITY: "Bit-Byte",
+}
+
+# TODO: support more flexible typing and string-to-var parsing/conversion
+def sort_rsvps_into_events(rsvps):
+    """ Takes in all confirmed rsvps and sorts them into types, currently hard coded. """
+    # Events in admin are currently in a readable format, must convert them to callable keys for Django template
+    sorted_events = dict.fromkeys(map_event_vars.keys())
+    for event_key, event_type in map_event_vars.items():
+        temp = []
+        for rsvp in rsvps.filter(event__event_type__type=event_type):
+            temp.append(rsvp.event)
+        sorted_events[event_key] = temp
+    return sorted_events
+
+
+def get_unconfirmed_events(rsvps):
+    unconfirmed_events = sort_rsvps_into_events(rsvps.filter(confirmed=False))
+
+    # We want to show all mandatory events, not just the events the candidate has RSVP'd to
+    # Get all mandatory events i.e. events with event type "Mandatory"
+    mandatory_events = Event.objects.filter(event_type__type=map_event_vars[settings.MANDATORY_EVENT])
+
+    # Initialize unconfirmed_events[settings.MANDATORY_EVENT]
+    if settings.MANDATORY_EVENT not in unconfirmed_events:
+        unconfirmed_events[settings.MANDATORY_EVENT] = []
+
+    for mandatory_event in mandatory_events:
+        # If no rsvps are found, add this mandatory event to the list of unconfirmed events
+        if rsvps.filter(event__id=mandatory_event.id).count() == 0:
+            unconfirmed_events[settings.MANDATORY_EVENT].append(mandatory_event)
+
+    return unconfirmed_events
+
+
+# TODO: increase flexibility by fetching event requirement count from database
+req_list = {
+    settings.MANDATORY_EVENT: 3,
+    settings.FUN_EVENT: 3,
+    settings.BIG_FUN_EVENT: 1,
+    settings.SERV_EVENT: 1,
+    settings.PRODEV_EVENT: 1,
+    settings.HANGOUT_EVENT: {
+        settings.HANGOUT_ATTRIBUTE_NAME: 1,
+        settings.CHALLENGE_ATTRIBUTE_NAME: 1,
+        settings.EITHER_ATTRIBUTE_NAME: 3,
```

Comment:
<img width="369" alt="Screen Shot 2020-10-06 at 12 48 21 PM" src="https://user-images.githubusercontent.com/46772636/95252728-3940b280-07d2-11eb-963b-76199e85cb56.png">

Just a suggestion - I think it would be helpful to have one extra section here for "General interactivity" or "Last interactivity" because it might look like they completed their requirements after only doing the two that are there.

Reply:
They'll only change to "completed" if they've completed all 3! See https://github.com/compserv/hknweb/blob/6962040bdeea65059ea147f8e10d96609117304f/hknweb/candidate/views.py#L374